Publish FLEDGE performance article (#3095)
GoogleChrome/developer.chrome.com: site/en/docs/privacy-sandbox/fledge-best-practices/index.md

---
layout: 'layouts/doc-post.njk'
title: 'FLEDGE latency best practices'
subhead: >

>   How to get the best performance from FLEDGE usage

Reviewer: ```suggestion
  Get the best performance from FLEDGE
```

description: >

>   Collection of best practices to improve FLEDGE auction latency

Reviewer: ```suggestion
  Review a collection of best practices to improve FLEDGE auction latency.
```

date: 2022-06-27
authors:
  - pauljensen
---

<!--lint disable no-smart-quotes-->

It's in everyone's best interest to make sure FLEDGE operates efficiently.

> * People browsing the web want sites to load quickly. This means FLEDGE must not overutilize limited device resources, like compute or network resources, necessary to load sites and their embedded ads.

Reviewer: Is it up to FLEDGE to not overuse resources? Or is it up to the site developer to build with FLEDGE efficiently, as to not overuse limited device resources?

Author: In the context of this article, it is up to the site developer to setup FLEDGE in an efficient way, since FLEDGE can be heavy in resources.

Reviewer: ```suggestion
* People browsing the web want sites to load quickly. This means developers should build with FLEDGE efficiently as to not overutilize limited device resources, like compute or network resources, that are necessary to load sites and their embedded ads.
```

Reviewer 2: This is a good suggestion, since the article is talking about best practice for the use of FLEDGE, not the implementation/design of FLEDGE.

* Publishers want their sites to load quickly, providing users an efficient and responsive experience. Publishers also want effective advertising to maximize their revenue.
* Advertisers and adtechs want their ads to display quickly to provide the greatest utility.

> This document seeks to outline some behaviors to ensure FLEDGE operates at maximum efficiency.

Reviewer: ```suggestion
This document outlines some best practices for FLEDGE implementation, to ensure your site operates at maximum efficiency.
```

## Buyer (bidder) best practices

To make sure you're optimizing for FLEDGE auction efficiency, follow these best practices

### Fewer interest group owners

To protect FLEDGE bidders in the same way that the browser protects different origins on the web using [site isolation](https://www.chromium.org/Home/chromium-security/site-isolation/), the browser uses expensive resources (like operating system processes) to protect individual interest group owners.

To minimize the expenditure of these very expensive resources, having the fewest number of interest group owners is crucial. Avoid having different interest groups owned by various subdomains, for example if `adtech.example` had interest groups owned by cats.`adtech.example` and `dogs.adtech.example`, the browser would likely use two separate processes to run their bidding scripts.

### Fewer interest groups bidding

> The browser must do significant setup and preparation before invoking a buyer's `generateBid()` script, such as setting up a new clean JavaScript execution environment, and parsing and loading the `generateBid()` code.

Reviewer: I'm confused by what "do significant setup and preparation" means. Can you clarify?

Author: I think in this case, the second sentence "such as setting up a new clean JavaScript execution environment, and parsing and loading the `generateBid()` code" describes that setup/preparation part.  If you think it needs more clarification, it may be worth reaching out to the original author of the article.

* Currently, FLEDGE supports seller-specified limits on the numbers of interest groups,and an API for buyers to specify the relative priority of their interest groups. This can be used to significantly reduce the number of bidding scripts to run.

There is ongoing discussion about how to filter the interest groups that participate in an auction. If you have ideas in this area, [file a GitHub issue](https://github.com/WICG/turtledove/issues/new) or contribute to [Issue 305](https://github.com/WICG/turtledove/issues/305).

There is ongoing discussion about how to reuse the JavaScript execution environment between auctions. Reusing it reduces the amount of time the browser must spend initializing them. You can follow along on [Issue 304](https://github.com/WICG/turtledove/issues/304) or [Issue 310](https://github.com/WICG/turtledove/issues/310), or [file a new GitHub issue](https://github.com/WICG/turtledove/issues/new).

### Reuse bidding scripts

> Interest groups sharing a bidding script can prevent the browser from having to download multiple scripts (which incurs extra network requests) and parse and compile them.

Reviewer: ```suggestion
Set up a shared bidding script for separate interest groups. This prevents the browser from having to download, parse, and compile multiple scripts (which incurs extra network requests).
```

### Reuse `trustedBiddingSignalsUrls`

Network latency and resource usage can be very significant. Fewer real-time trusted bidding signals fetches can help reduce this latency.

> As the explainer notes, the trusted bidding [signal fetches can be combined](https://github.com/WICG/turtledove/blob/main/FLEDGE.md#11-joining-interest-groups) when the `trustedBiddingSignalsUrl` is reused amongst multiple interest groups, so be sure to use the same `trustedBiddingSignalsUrl` between interest groups when possible.

Reviewer: ```suggestion
As the explainer notes, the [trusted bidding signal fetches can be combined](https://github.com/WICG/turtledove/blob/main/FLEDGE.md#11-joining-interest-groups) when the `trustedBiddingSignalsUrl` is reused amongst multiple interest groups, so be sure to use the same `trustedBiddingSignalsUrl` between interest groups when possible.
```

### Smaller real-time trusted bidding signals fetches

Network latency can be very significant, and this is directly impacted by how much data is transferred during the real-time trusted bidding signal fetches. Prefer storing ad-specific or interest-group-specific data in the interest group, rather than on the real-time trusted bidding signal server. Reserve the real-time trusted bidding signal data for only those truly real-time signals, like campaign budgeting or kill-switches.

Any signal that can be updated on a daily or longer basis should be stored in the interest group and updated via the daily updates.

## Seller best practices

Make sure you're monitoring and optimizing for FLEDGE auction efficiency.

### Monitor your auctions

Collect metrics on your auctions:

> * As discussed above, the number of interest group owners and interest groups bidding (both visible metrics to `scoreAd()`) have a significant effect on auction performance and are worth monitoring.

Reviewer: ```suggestion
* The number of interest group owners and interest groups bidding significantly impacts auction performance and should be monitored. The metrics are visible in `scoreAd()`.
```

Reviewer: Someone may not read the above if they're only interested in the seller side. That said, you could link above to the corresponding issue.
